### ISSUE TITLE
Incorrect scan numbers on filtered files

### ISSUE THREAD

**Issue by chambm** [Nesvilab/MSFragger#196]
I have a test file that is filtered from 100 to 110 minutes into my run. When I run MSFragger 3.4 on it, I get correct native_ids in the pepXML output, but the scan numbers in the pepXML, TSV, and PIN outputs all start at 1 instead of 4610. This is also a problem with a file filtered to just MS2s (the scan numbers would be non-contiguous).

**Comment by fcyu** [comment]
Hi Matt,

You are correct. We use `index + 1` as the scan number, rather than the `spectrum title` since not all mzML files have it. Do you have any suggestions, which can cover most mzML flavors?

Thanks,

Fengchao

**Comment by chambm** [comment]
It's true you can't get a real scan number for a WIFF or Waters RAW. That's why I don't really recommend scan numbers at all. But since tons of people persist in using them, I use this function to TRY to get a real scan number:
```
string translateNativeIDToScanNumber(CVID nativeIdFormat, const string& id)
{
    switch (nativeIdFormat)
    {
        case MS_spectrum_identifier_nativeID_format: // mzData
            return value(id, "spectrum");

        case MS_multiple_peak_list_nativeID_format: // MGF
            return value(id, "index");

        case MS_Agilent_MassHunter_nativeID_format:
            return value(id, "scanId");

        case MS_Thermo_nativeID_format:
            // conversion from Thermo nativeIDs assumes default controller information
            if (id.find("controllerType=0 controllerNumber=1") != 0)
                return "";

            // fall through to get scan

        case MS_Bruker_Agilent_YEP_nativeID_format:
        case MS_Bruker_BAF_nativeID_format:
        case MS_scan_number_only_nativeID_format:
            return value(id, "scan");

        default:
            if (bal::starts_with(id, "scan=")) return value(id, "scan");
            else if (bal::starts_with(id, "index=")) return value(id, "index");
            return "";
    }
}
```
An empty string returned means a scan number couldn't be extracted from the nativeID. In those cases I'll use index+1 as you do.

**Comment by chambm** [comment]
I'm sure the function could be simplified if you don't know the nativeIdFormat at a minimal cost to robustness.

**Comment by fcyu** [comment]
Hi Matt,

Thank you very much for your suggestion. I think I get the point of your code. It tries to match the spectrum ID, and extract the scan number. If there is no such match, using `index + 1`. I will implement this in the next MSFragger release.

Best,

Fengchao

**Comment by fcyu** [comment]
Hi Matt @chambm ,

I just found that using the scan number from `controllerType=0 controllerNumber=1 scan=2` for the demultiplexed data would cause issues. 

For examples, two demultiplexed MS/MS scans have `controllerType=0 controllerNumber=1 scan=3 demux=0` and `controllerType=0 controllerNumber=1 scan=3 demux=1`. They basically have the same "scan number" if using `scan=3`. Any suggestions for this case?

Thanks,

Fengchao

**Comment by chambm** [comment]
By convention, if a spectrum is synthesized from multiple others or vice
versa (one spectrum creates multiple synthesized ones), the id starts with
"merged=(index)". Then you can put whatever you want after it. So to get a
single scan number from that, index makes the most sense.

**Comment by fcyu** [comment]
Fixed in MSFragger 3.5.

Best,

Fengchao
